### DISCUSSION TITLE
'type' folder layout changed and unable to import type for ethers.

### DISCUSSION THREAD

**Discussion by code-brewer** [PaulRBerg/hardhat-template#169] [Q&A]
I use this template, it compile and test ok without modification.

After that, I installed '"@openzeppelin/contracts": "^4.9.2"' , and add my empty contact that inherit from openzeppelin's Ownable

```
contract DataStore is Ownable{
}
```
Then, the vscode IDE prompt unable to import types, such as `Cannot find module 'hardhat' or its corresponding type declarations.`

I notice that the 'type' folder layout changed.

Origianl was:
```bash
types
├── common.ts
├── factories
│   ├── Greeter__factory.ts
│   └── index.ts
├── Greeter.ts
├── hardhat.d.ts
└── index.ts
```

now the layout is:
```
types
├── @openzeppelin
│   ├── contracts
│   │   ├── access
│   │   │   ├── Ownable.ts
│   │   │   └── index.ts
│   │   └── index.ts
│   └── index.ts
├── common.ts
├── contracts
│   ├── DataStore.ts
│   ├── Greeter.ts
│   └── index.ts
├── factories
│   ├── @openzeppelin
│   │   ├── contracts
│   │   │   ├── access
│   │   │   │   ├── Ownable__factory.ts
│   │   │   │   └── index.ts
│   │   │   └── index.ts
│   │   └── index.ts
│   ├── contracts
│   │   ├── DataStore__factory.ts
│   │   ├── Greeter__factory.ts
│   │   └── index.ts
│   └── index.ts
├── hardhat.d.ts
└── index.ts
```

After fix Greeter/Greeter__factory import path , as:
```
import type { Greeter } from "../../types/contracts/Greeter";
import type { Greeter__factory } from "../../types/factories/contracts/Greeter__factory";
```
there still error for `import { ethers } from "hardhat";`  
![image](https://github.com/PaulRBerg/hardhat-template/assets/9084210/d833d164-3e9a-43f7-9f8c-7fa3e687cda1)

So, I thought about two question:
1) why the 'type' folder layout changed ? is there configuration parameter to keep original layout?
2) how to fix type import problem for 'ethers'?

Thanks.

**Comment by PaulRBerg** [comment]
Hello @code-brewer! Thanks for taking the time to write such a detailed question.

Unfortunately, I am no longer using this template on a regular basis (I am using [Foundry](https://github.com/PaulRBerg/foundry-template) now), so I don't know how to answer your question. You might have better luck looking for help in the [Hardhat Discord server](https://discord.com/invite/TETZs2KK4k).

**Comment by SvenMeyer** [comment]
@code-brewer thanks for decumenting what I also experienced.

I just created a new project with the template from hardhat. That is pretty minimalistic in its configuration.

hardhat.config.ts
```javascript
import { HardhatUserConfig } from "hardhat/config";
import "@nomicfoundation/hardhat-toolbox";

const config: HardhatUserConfig = {
  solidity: "0.8.19",
};

export default config;
```

...  and looks like it does not show that (new ?) behavior.

![image](https://github.com/PaulRBerg/hardhat-template/assets/25609/18936e02-55a4-4b77-8c2c-2e6832c0e45f)


@PaulRBerg maybe behaviours is (now) different if `paths` are specified explicitly ? :thinking: 
Do we even still need that ?
... and another question is, what behaviour do we even want ?

**Comment by SvenMeyer** [comment]
Update: even more confusing, I started with a fresh now template yesterday, replaced Greeter with my contract ... and it did not put the *.ts files in subfolders.

Then I added another contract, also straight into the `contracts` folder. Now this ends up in subdirectories like described by @code-brewer above, but the first contract not.
Doing a fresh re-creation of type files with `pnpm clean` also does not change anything :thinking: 

```
$ tree  -L 3 types/
types/
├── common.ts
├── contracts
│   ├── AccessControlToken.sol
│   │   ├── index.ts
│   │   └── TestERC20.ts
│   ├── index.ts
│   └── FirstContract.ts
├── factories
│   ├── contracts
│   │   ├── AccessControlToken.sol
│   │   ├── index.ts
│   │   └── FirstContract__factory.ts
│   ├── index.ts
│   └── @openzeppelin
│       ├── contracts
│       └── index.ts
├── hardhat.d.ts
├── index.ts
└── @openzeppelin
    ├── contracts
    │   ├── access
    │   ├── index.ts
    │   ├── interfaces
    │   ├── security
    │   ├── token
    │   └── utils
    └── index.ts
```

The standard hardhat template does not show the behaviour.

```
$ tree  -L 3 typechain-types/
typechain-types/
├── common.ts
├── factories
│   ├── index.ts
│   ├── LockCopy__factory.ts
│   └── Lock__factory.ts
├── hardhat.d.ts
├── index.ts
├── LockCopy.ts
└── Lock.ts

```

I looked if I could find some configuration somewhere which might cause the different behaviour, but could not find anything.
